Index: eglib/src/goutput.c
===================================================================
--- eglib/src/goutput.c	(revision 149256)
+++ eglib/src/goutput.c	(working copy)
@@ -28,6 +28,10 @@
 #include <stdlib.h>
 #include <glib.h>
 
+#ifdef PLATFORM_ANDROID
+#include <android/log.h>
+#endif
+
 /* The current fatal levels, error is always fatal */
 static GLogLevelFlags fatal = G_LOG_LEVEL_ERROR;
 
@@ -39,6 +43,10 @@
 	va_start (args, format);
 	vprintf (format, args);
 	va_end (args);
+	
+#ifdef PLATFORM_ANDROID
+    __android_log_vprint(ANDROID_LOG_INFO, "Mono", format, args);
+#endif
 }
 
 GLogLevelFlags
