Index: eglib/src/glib.h
===================================================================
--- eglib/src/glib.h	(revision 147279)
+++ eglib/src/glib.h	(working copy)
@@ -496,7 +496,9 @@
 	G_LOG_LEVEL_MASK              = ~(G_LOG_FLAG_RECURSION | G_LOG_FLAG_FATAL)
 } GLogLevelFlags;
 
-void           g_print                (const gchar *format, ...);
+//void           g_print                (const gchar *format, ...);
+#define LOG_TAG "Mono"
+#define g_print LOGI
 GLogLevelFlags g_log_set_always_fatal (GLogLevelFlags fatal_mask);
 GLogLevelFlags g_log_set_fatal_mask   (const gchar *log_domain, GLogLevelFlags fatal_mask);
 void           g_logv                 (const gchar *log_domain, GLogLevelFlags log_level, const gchar *format, va_list args);
Index: mono/mini/mini.c
===================================================================
--- mono/mini/mini.c	(revision 147279)
+++ mono/mini/mini.c	(working copy)
@@ -5007,6 +5007,8 @@
 	domain->runtime_info = NULL;
 }
 
+void register_icalls ();
+
 MonoDomain *
 mini_init (const char *filename, const char *runtime_version)
 {
@@ -5183,6 +5185,7 @@
 	mono_add_internal_call ("Mono.Runtime::mono_runtime_install_handlers", 
 				mono_runtime_install_handlers);
 
+	register_icalls ();
 
 	create_helper_signature ();
 
