Index: eglib/src/goutput.c
===================================================================
--- eglib/src/goutput.c	(revision 147505)
+++ eglib/src/goutput.c	(working copy)
@@ -28,6 +28,10 @@
 #include <stdlib.h>
 #include <glib.h>
 
+#ifdef PLATFORM_ANDROID
+#include <android/log.h>
+#endif
+
 /* The current fatal levels, error is always fatal */
 static GLogLevelFlags fatal = G_LOG_LEVEL_ERROR;
 
@@ -39,6 +43,10 @@
 	va_start (args, format);
 	vprintf (format, args);
 	va_end (args);
+	
+#ifdef PLATFORM_ANDROID
+    __android_log_vprint(ANDROID_LOG_INFO, "Mono", format, args);
+#endif
 }
 
 GLogLevelFlags
Index: mono/mini/mini.c
===================================================================
--- mono/mini/mini.c	(revision 147505)
+++ mono/mini/mini.c	(working copy)
@@ -5007,6 +5007,8 @@
 	domain->runtime_info = NULL;
 }
 
+void register_icalls ();
+
 MonoDomain *
 mini_init (const char *filename, const char *runtime_version)
 {
@@ -5183,6 +5185,7 @@
 	mono_add_internal_call ("Mono.Runtime::mono_runtime_install_handlers", 
 				mono_runtime_install_handlers);
 
+	register_icalls ();
 
 	create_helper_signature ();
 
