diff --git a/eglib/src/goutput.c b/eglib/src/goutput.c
index bce2f0a..4ab6416 100644
--- a/eglib/src/goutput.c
+++ b/eglib/src/goutput.c
@@ -54,6 +54,7 @@ out_vfprintf (FILE *ignore, const gchar *format, va_list args)
 {
 	/* TODO: provide a proper app name */
 	__android_log_vprint (ANDROID_LOG_ERROR, "mono", format, args);
+	vfprintf (ignore, format, args);
 }
 #else
 static void 
diff --git a/mono/mini/mini.c b/mono/mini/mini.c
index 71d7f26..7375b64 100644
--- a/mono/mini/mini.c
+++ b/mono/mini/mini.c
@@ -83,6 +83,8 @@
 #define MINI_FAST_TLS_DECLARE(x) static pthread_key_t x;
 #define MINI_HAVE_FAST_TLS
 #define MINI_THREAD_VAR_OFFSET(x,y) y = (gint32) x
+#elif (defined(PLATFORM_ANDROID))
+#define MINI_THREAD_VAR_OFFSET(x,y) MONO_THREAD_VAR_OFFSET(x,y)
 #endif
 
 static gpointer mono_jit_compile_method_with_opt (MonoMethod *method, guint32 opt, MonoException **ex);
